File: clients/java/src/main/java/com/orisunlabs/orisun/client/TokenCache.java
package com.orisunlabs.orisun.client;

import io.grpc.Metadata;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Token cache for storing and managing authentication tokens
 */
public class TokenCache {
    private final AtomicReference<String> cachedToken = new AtomicReference<>();
    private final Logger logger;
    
    public TokenCache(Logger logger) {
        this.logger = logger;
    }
    
    /**
     * Store a token in the cache
     * @param token The token to cache
     */
    public void cacheToken(String token) {
        if (token != null && !token.trim().isEmpty()) {
            cachedToken.set(token);
            logger.debug("Cached authentication token");
        }
    }
    
    /**
     * Get the cached token
     * @return The cached token, or null if no token is cached
     */
    public String getCachedToken() {
        return cachedToken.get();
    }
    
    /**
     * Check if a token is cached
     * @return true if a token is cached
     */
    public boolean hasToken() {
        return cachedToken.get() != null;
    }
    
    /**
     * Clear the cached token
     */
    public void clearToken() {
        cachedToken.set(null);
        logger.debug("Cleared cached authentication token");
    }
    
    /**
     * Extract token from response metadata and cache it
     * @param headers The response metadata
     */
    public void extractAndCacheToken(Metadata headers) {
        if (headers != null) {
            Metadata.Key<String> tokenKey = Metadata.Key.of("x-auth-token", Metadata.ASCII_STRING_MARSHALLER);
            String token = headers.get(tokenKey);
            
            if (token != null && !token.trim().isEmpty()) {
                cacheToken(token);
                logger.debug("Extracted and cached token from response headers");
            }
        }
    }
    
    /**
     * Create metadata with authentication (cached token or basic auth)
     * @param basicAuthCredentials Basic auth credentials to use if no token is cached
     * @return Metadata with authentication headers
     */
    public Metadata createAuthMetadata(String basicAuthCredentials) {
        Metadata metadata = new Metadata();
        
        String token = getCachedToken();
        if (token != null) {
            metadata.put(Metadata.Key.of("x-auth-token", Metadata.ASCII_STRING_MARSHALLER), token);
            logger.debug("Using cached authentication token");
        } else if (basicAuthCredentials != null) {
            metadata.put(Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER), basicAuthCredentials);
            logger.debug("Using basic authentication");
        }
        
        return metadata;
    }
}

File: clients/go/token_cache.go
package orisun

import (
	"encoding/base64"
	"sync"

	"google.golang.org/grpc/metadata"
)

// TokenCache for storing and managing authentication tokens
type TokenCache struct {
	cachedToken *string
	logger       Logger
	mu           sync.RWMutex
}

// NewTokenCache creates a new TokenCache with the given logger
func NewTokenCache(logger Logger) *TokenCache {
	if logger == nil {
		logger = NewNoOpLogger()
	}
	return &TokenCache{
		logger: logger,
	}
}

// CacheToken stores a token in the cache
func (tc *TokenCache) CacheToken(token string) {
	if token == "" {
		return
	}
	
	tc.mu.Lock()
	defer tc.mu.Unlock()
	
	tc.cachedToken = &token
	tc.logger.Debug("Cached authentication token")
}

// GetCachedToken returns the cached token, or empty string if no token is cached
func (tc *TokenCache) GetCachedToken() string {
	tc.mu.RLock()
	defer tc.mu.RUnlock()
	
	if tc.cachedToken == nil {
		return ""
	}
	return *tc.cachedToken
}

// HasToken returns true if a token is cached
func (tc *TokenCache) HasToken() bool {
	tc.mu.RLock()
	defer tc.mu.RUnlock()
	
	return tc.cachedToken != nil
}

// ClearToken clears the cached token
func (tc *TokenCache) ClearToken() {
	tc.mu.Lock()
	defer tc.mu.Unlock()
	
	tc.cachedToken = nil
	tc.logger.Debug("Cleared cached authentication token")
}

// ExtractAndCacheToken extracts token from response metadata and caches it
func (tc *TokenCache) ExtractAndCacheToken(md metadata.MD) {
	if md == nil {
		return
	}
	
	values := md.Get("x-auth-token")
	if len(values) > 0 {
		token := values[0]
		if token != "" {
			tc.CacheToken(token)
			tc.logger.Debug("Extracted and cached token from response headers")
		}
	}
}

// CreateAuthMetadata creates metadata with authentication (cached token or basic auth)
func (tc *TokenCache) CreateAuthMetadata(basicAuthCredentials string) metadata.MD {
	md := metadata.New(nil)
	
	token := tc.GetCachedToken()
	if token != "" {
		md.Set("x-auth-token", token)
		tc.logger.Debug("Using cached authentication token")
	} else if basicAuthCredentials != "" {
		md.Set("Authorization", basicAuthCredentials)
		tc.logger.Debug("Using basic authentication")
	}
	
	return md
}

// CreateBasicAuthCredentials creates basic authentication credentials from username and password
func CreateBasicAuthCredentials(username, password string) string {
	if username == "" || password == "" {
		return ""
	}
	
	auth := username + ":" + password
	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
}